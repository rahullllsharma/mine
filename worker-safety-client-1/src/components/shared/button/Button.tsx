import type { IconName } from "@urbint/silica";
import type {
  ButtonHTMLAttributes,
  ForwardedRef,
  HTMLAttributes,
  ReactElement,
  ReactNode,
} from "react";
import { Icon } from "@urbint/silica";
import React, { forwardRef } from "react";

import cx from "classnames";
import { IconSpinner } from "@/components/iconSpinner";

export type ButtonProps = Omit<
  HTMLAttributes<HTMLButtonElement>,
  "children"
> & {
  controlStateClass?: string;
  iconStart?: IconName;
  iconEnd?: IconName;
  label?: string | ReactElement;
  disabled?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  type?: ButtonHTMLAttributes<HTMLButtonElement>["type"];
  loading?: boolean;
  labelClassName?: string;
  iconStartClassName?: string;
  children?: ReactNode;
  iconEndClassName?: string;
};

function Button(
  {
    className,
    controlStateClass,
    iconStart,
    iconEnd,
    label,
    type = "button",
    loading = false,
    iconStartClassName = "",
    labelClassName = "",
    children,
    iconEndClassName = "",
    ...tailProps
  }: ButtonProps,
  ref: ForwardedRef<HTMLButtonElement>
): JSX.Element {
  const enableControlClassName = tailProps.disabled
    ? ""
    : "hover:bg-neutral-light-16 focus:bg-neutral-shade-7 active:bg-neutral-shade-7";

  return (
    <button
      className={`text-center truncate disabled:opacity-38 disabled:cursor-not-allowed ${className}`}
      type={type}
      ref={ref}
      {...tailProps}
      disabled={tailProps.disabled || loading}
    >
      <div
        className={`flex items-center justify-center h-full ${enableControlClassName} ${controlStateClass}`}
      >
        {iconStart && <Icon className={iconStartClassName} name={iconStart} />}
        {label && (
          <span className={cx("mx-1 truncate", labelClassName)}>{label}</span>
        )}
        {children}
        {loading && <IconSpinner />}
        {iconEnd && !loading && (
          <Icon name={iconEnd} className={iconEndClassName} />
        )}
      </div>
    </button>
  );
}

export default forwardRef(Button);
