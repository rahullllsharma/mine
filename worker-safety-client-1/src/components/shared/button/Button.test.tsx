import { render, screen } from "@testing-library/react";
import React from "react";
import Button from "./Button";

describe("ButtonPrimary", () => {
  it("should be in the document", () => {
    const { getByRole } = render(<Button label="ok" />);

    expect(getByRole("button")).toBeInTheDocument();
  });

  describe("when is disabled", () => {
    it("should have the attribute disabled", () => {
      const { getByRole } = render(<Button disabled label="ok" />);
      expect(getByRole("button")).toBeDisabled();
    });
  });

  describe("when it's loading", () => {
    it("should show the loading spinner", () => {
      const { asFragment } = render(<Button loading label="ok" />);

      expect(screen.getByRole("button")).toBeDisabled();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("when contain icons", () => {
    describe("when is passed a left icon", () => {
      it("should contain an icon as first child", () => {
        const { asFragment } = render(
          <Button label="ok" iconStart="chevron_duo_up" />
        );
        expect(asFragment()).toMatchSnapshot();
      });
    });
    describe("when is passed a right icon", () => {
      it("should contain an icon as last child", () => {
        const { asFragment } = render(
          <Button label="ok" iconEnd="chevron_duo_up" />
        );
        expect(asFragment()).toMatchSnapshot();
      });
    });
    describe("when is passed a left and right icon", () => {
      it("should contain both icons", () => {
        const { asFragment } = render(
          <Button
            label="ok"
            iconStart="chevron_duo_up"
            iconEnd="chevron_duo_up"
          />
        );
        expect(asFragment()).toMatchSnapshot();
      });
    });
  });
});
