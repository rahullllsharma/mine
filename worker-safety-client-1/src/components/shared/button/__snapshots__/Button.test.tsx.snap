// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonPrimary when contain icons when is passed a left and right icon should contain both icons 1`] = `
<DocumentFragment>
  <button
    class="text-center truncate disabled:opacity-38 disabled:cursor-not-allowed undefined"
    type="button"
  >
    <div
      class="flex items-center justify-center h-full hover:bg-neutral-light-16 focus:bg-neutral-shade-7 active:bg-neutral-shade-7 undefined"
    >
      <i
        aria-hidden="true"
        class="ci-chevron_duo_up"
      />
      <span
        class="mx-1 truncate"
      >
        ok
      </span>
      <i
        aria-hidden="true"
        class="ci-chevron_duo_up"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`ButtonPrimary when contain icons when is passed a left icon should contain an icon as first child 1`] = `
<DocumentFragment>
  <button
    class="text-center truncate disabled:opacity-38 disabled:cursor-not-allowed undefined"
    type="button"
  >
    <div
      class="flex items-center justify-center h-full hover:bg-neutral-light-16 focus:bg-neutral-shade-7 active:bg-neutral-shade-7 undefined"
    >
      <i
        aria-hidden="true"
        class="ci-chevron_duo_up"
      />
      <span
        class="mx-1 truncate"
      >
        ok
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`ButtonPrimary when contain icons when is passed a right icon should contain an icon as last child 1`] = `
<DocumentFragment>
  <button
    class="text-center truncate disabled:opacity-38 disabled:cursor-not-allowed undefined"
    type="button"
  >
    <div
      class="flex items-center justify-center h-full hover:bg-neutral-light-16 focus:bg-neutral-shade-7 active:bg-neutral-shade-7 undefined"
    >
      <span
        class="mx-1 truncate"
      >
        ok
      </span>
      <i
        aria-hidden="true"
        class="ci-chevron_duo_up"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`ButtonPrimary when it's loading should show the loading spinner 1`] = `
<DocumentFragment>
  <button
    class="text-center truncate disabled:opacity-38 disabled:cursor-not-allowed undefined"
    disabled=""
    type="button"
  >
    <div
      class="flex items-center justify-center h-full hover:bg-neutral-light-16 focus:bg-neutral-shade-7 active:bg-neutral-shade-7 undefined"
    >
      <span
        class="mx-1 truncate"
      >
        ok
      </span>
      <i
        aria-hidden="true"
        class="animate-spin"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.75 17.25c-.615.615-1.626.622-2.125-.091a9 9 0 01-.446-.706c-.43-.756.008-1.666.828-1.958l.263-.095c.674-.24 1.404.123 1.786.727.382.605.397 1.42-.108 1.925l-.198.198zM10.144 4.81c-.217-.842.29-1.715 1.158-1.783A9 9 0 118.848 20.43c-.814-.305-1.061-1.285-.62-2.034.442-.75 1.408-.975 2.248-.748a5.85 5.85 0 101.621-11.497c-.87-.015-1.735-.498-1.953-1.34z"
          />
        </svg>
      </i>
    </div>
  </button>
</DocumentFragment>
`;
